### PR TITLE
Generate matrix with item-piechart directive

### LINKED ISSUE #368: item-matrix should support the same categories as item-piechart
In my project, I have the following instance of `item-piechart`:
```
.. item-piechart:: Unit requirement pie chart
    :id_set: RQT- TC- TR-
    :label_set: UNCOVERED, COVERED, EXECUTED
    :sourcetype: validated_by
    :targettype: tested_by
    :scope: UNIT
    :result: FAILED, PASSED
    :colors: #b2c4cb #eea320 #b2c4cb #db4140 #65bba9
```

Which gives the following pie chart:
image

I also have an `item-matrix` with the same requirements as the pie chart:
```
.. item-matrix:: Unit requirement matrix
    :source: RQT-
    :intermediate: TC-
    :target: TR-
    :sourcetitle: Requirement
    :intermediatetitle: Testcase
    :targettitle: Test record
    :type: validated_by | tested_by
    :scope: UNIT
    :targetcolumns: result
    :coveredintermediates:
    :splitintermediates:
```

I would like to split my matrix into the same five subsections (or into five submatrices) as the pie chart. However, the `item-matrix` directive currently only provides the `onlyuncovered` and `onlycovered` filters. The first filter gives me the sum of UNCOVERED and COVERED, while the second gives me the rest.

### PR DESCRIPTION
**Pull request by JasperCraeghs**
- [x] rough implementation
- [x] tweak table layout: group by result with result in rowspan/subheader
- [x] clean up code
- [x] testing
- [x] documentation

Closes #368

FYI @gcrabbe

### CODE REVIEW

**Comment by JasperCraeghs** [comment]
The case-insentivity of the labels must be taken into account. The following configuration doesn't produce any items under the subheader `ERROR`:

```
    :result: ERROR, fail, pass
    :matrix:
```

File: doc/requirements.rst [hunk 1]
```diff
@@ -156,20 +157,15 @@ Test coverage
     :label_set: not covered, covered
     :colors: orange c b
     :stats:
+    :matrix:
 
-.. item-piechart:: Test coverage chart targeting UTEST and ITEST
+.. item-piechart:: Test coverage chart with test results, based on the :targettype: option
```
Comment (Letme) [inline doc/requirements.rst:162]: To keep aligned with `:result: attrubute` below

```suggestion
.. item-piechart:: Test coverage chart with test results, based on the :targettype: attribute
```

File: doc/requirements.rst [hunk 2]
```diff
@@ -213,6 +202,16 @@ Test coverage
     :splitsourcetype:
     :stats:
 
+.. item-piechart:: Test coverage chart targeting ITEST only
```
Comment (Letme) [inline doc/requirements.rst:205]:
```suggestion
.. item-piechart:: Test coverage chart targeting ITEST only with matrix attribute containing filtered labels
```

File: mlx/traceability/directives/item_pie_chart_directive.py [hunk 3]
```diff
@@ -358,6 +396,76 @@ def _get_explode_values(labels, label_set):
...
+            match (Match): The corresponding Match instance
+            app (sphinx.application.Sphinx): Sphinx application object
+
+        Returns:
+            list: List of rows to add to the table body
+        """
+        rows = []
+        source_row = nodes.row()
+        source_row += self._create_cell_for_items([source], app, morerows=max(0, len(match.targets)-1))
+        if match.targets:
```
Comment (Letme) [inline mlx/traceability/directives/item_pie_chart_directive.py:454]: This if statement is tricky. If it does not exist, above len returns 0, so you get -1 in the attribute of the max function, but if it is 1 or more then you enter this if statement. 

Plus additionally in the if statement you depend that `match.targets` is connected to `match.targets_iter` so maybe more correctly here it should be:

```suggestion
        if match.targets_iter:
```

File: mlx/traceability/directives/item_pie_chart_directive.py [hunk 4]
```diff
@@ -179,8 +218,9 @@ def _match_covered(self, top_source_id, nested_source_item, relationship):
         Args:
             top_source_id (str): Identifier of the top source item, e.g. requirement identifier.
             nested_source_item (None/TraceableItem): Nested traceable item to be used as a source for looping through
-                its relationships, e.g. a test item. If None, only the given `relationship` is taken into account.
+                its relationships, e.g. a test item.
```
Comment (Letme) [inline mlx/traceability/directives/item_pie_chart_directive.py:221]: Here the `None` statement still holds true, but it changes a bit considering the `consider_nested_targets` so maybe better:

```suggestion
                its relationships, e.g. a test item. If None, the consider_nested_targets could configure the matches `label`
```

**Review by Letme — CHANGES_REQUESTED** [review]

File: doc/requirements.rst (hunk 1 above)
Comment (JasperCraeghs) [reply doc/requirements.rst:162]: No. The `:result:` attribute is the `:<<attribute>>:` option. "targettype" is not a traceable attribute, it's an option of this directive.

File: mlx/traceability/directives/item_pie_chart_directive.py (hunk 3 above)
Comment (JasperCraeghs) [reply mlx/traceability/directives/item_pie_chart_directive.py:454]: I don't think it's tricky. I'm quite happy with the code. The doc of `max` states `With a single iterable argument, return its biggest item`

If the element is indeed `-1`, it'll result in `max(0, -1)`, which is `morerows=0`, which is what I want.

`targets_iter` is a generator, which is always `True` when evaluated as a `bool`, so must not be used in this if-statement like you suggested.

Can you explain what you find tricky please?

File: mlx/traceability/directives/item_pie_chart_directive.py (hunk 4 above)
Comment (JasperCraeghs) [reply mlx/traceability/directives/item_pie_chart_directive.py:221]: Huh? `consider_nested_targets` has no relation to `has_nested_target == None` due to `or` in `if not has_nested_target or not consider_nested_targets:`

The documentation still makes complete sense to me as it is now.

Comment (Letme) [reply mlx/traceability/directives/item_pie_chart_directive.py:221]: I stated it wrongly due to that `not`. You are correct.

File: mlx/traceability/directives/item_pie_chart_directive.py (hunk 3 above)
Comment (Letme) [reply mlx/traceability/directives/item_pie_chart_directive.py:454]: What I find tricky is if `match.targets does` not exist, the `len` function will operate on non-existing variable (assumingly return 0). Then this if statement here, which protects against non-existing variable is should be higher up, to protect against that, but now it seems like it is here just to work when `match.targets > 1`, so  you enter this statement when there is something inside, although `len` is already telling you there is 0 from above. Maybe this is normal for python and I am overthinking.

So `targets_iter` is produced only when `targets` exists? Then these two are connected so if statement makes sense (did not know that `_iter` is part of python syntax).

File: doc/requirements.rst (hunk 2 above)
Comment (JasperCraeghs) [reply doc/requirements.rst]: Improved title in 682c6222802b68dd65013926b5f7b4a04ebc0b4d

File: mlx/traceability/directives/item_pie_chart_directive.py (hunk 3 above)
Comment (Letme) [reply mlx/traceability/directives/item_pie_chart_directive.py:456]: Ok, thanks for explanation.

**Review by Letme — APPROVED** [review]

**Pull request merged**